src/ripple/consensus/Consensus.h

```diff
@@ -1292,13 +1514,80 @@ Consensus<Adaptor>::phaseEstablish()
     prevRoundTime_ = result_->roundTime.read();
     phase_ = ConsensusPhase::accepted;
     JLOG(j_.debug()) << "transitioned to ConsensusPhase::accepted";
+
+    std::optional<std::pair<
+        typename Adaptor::CanonicalTxSet_t,
+        typename Adaptor::Ledger_t>>
+        txsBuilt;
+    // Track time spent retrying new ledger validation.
+    std::optional<std::chrono::time_point<std::chrono::steady_clock>>
+        startDelay;
+    // Amount of time to pause checking for ledger to become validated.
+    static auto const validationWait = std::chrono::milliseconds(100);
+
+    // Building the new ledger is time-consuming and safe to not lock, but
+    // the rest of the logic below needs to be locked, until
+    // finishing (onAccept).
+    std::unique_lock<std::recursive_mutex> lock(adaptor_.peekMutex());
+    do
+    {
+        if (txsBuilt)
+        {
+            if (!startDelay)
+                startDelay = std::chrono::steady_clock::now();
+
+            // Only send a single validation per round.
+            adaptor_.clearValidating();
+            // Check if a better proposal has been shared by the network.
+            auto prevProposal = result_->position;
+            updateOurPositions(false);
+
+            if (prevProposal == result_->position)
+            {
+                JLOG(j_.debug())
+                    << "old and new positions "
+                       "match: "
+                    << prevProposal.position() << " delay so far "
+                    << std::chrono::duration_cast<std::chrono::milliseconds>(
+                           std::chrono::steady_clock::now() - *startDelay)
+                           .count()
+                    << "ms. pausing";
+                adaptor_.getLedgerMaster().waitForValidated(validationWait);
+                continue;
+            }
+            JLOG(j_.debug()) << "retrying buildAndValidate with "
+                                "new position: "
+                             << result_->position.position();
+        }
+        lock.unlock();
+
+        // This is time-consuming and safe to not have under mutex.
+        txsBuilt = adaptor_.buildAndValidate(
+            *result_,
+            previousLedger_,
+            closeResolution_,
+            mode_.get(),
+            getJson(true));
+        lock.lock();
+    } while (adaptor_.retryAccept(txsBuilt->second, startDelay));
+
+    if (startDelay)
+    {
+        auto const delay =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::steady_clock::now() - *startDelay);
+        JLOG(j_.debug()) << "validationDelay will be " << delay.count() << "ms";
+        adaptor_.setValidationDelay(delay);
+    }
+
+    lock.unlock();
+
     adaptor_.onAccept(
         *result_,
-        previousLedger_,
-        closeResolution_,
         rawCloseTimes_,
         mode_.get(),
-        getJson(true));
+        getJson(true),
+        std::move(*txsBuilt));
```

Comment:
`result_` is not always set when I run the `Consensus` unit test. It asserts in Windows. Since `phaseEstablish` starts with `assert(result_);`, I can only assume it's getting prematurely cleared somewhere.

Here's the stack
```
 	rippled.exe!std::optional<ripple::ConsensusResult<ripple::test::csf::Peer>>::operator*() Line 359	C++
>	rippled.exe!ripple::Consensus<ripple::test::csf::Peer>::phaseEstablish() Line 1587	C++
 	rippled.exe!ripple::Consensus<ripple::test::csf::Peer>::timerEntry(const std::chrono::time_point<ripple::NetClock,std::chrono::duration<unsigned int,std::ratio<1,1>>> & now) Line 986	C++
 	rippled.exe!ripple::test::csf::Peer::timerEntry() Line 932	C++
 	rippled.exe!`ripple::test::csf::Peer::timerEntry'::`4'::<lambda_1>::operator()() Line 933	C++
 	rippled.exe!ripple::test::csf::Scheduler::event_impl<`ripple::test::csf::Peer::timerEntry'::`4'::<lambda_1>>::operator()() Line 106	C++
 	rippled.exe!ripple::test::csf::Scheduler::step_one() Line 398	C++
 	rippled.exe!ripple::test::csf::Scheduler::step() Line 408	C++
 	rippled.exe!ripple::test::Consensus_test::testStandalone() Line 135	C++
 	rippled.exe!ripple::test::Consensus_test::run() Line 1058	C++
 	rippled.exe!beast::unit_test::suite::run<void>(beast::unit_test::runner & r) Line 589	C++
 	rippled.exe!beast::unit_test::suite::operator()<void>(beast::unit_test::runner & r) Line 429	C++
 	rippled.exe!`beast::unit_test::make_suite_info<ripple::test::Consensus_test>'::`2'::<lambda_1>::operator()(beast::unit_test::runner & r) Line 120	C++
 	[External Code]	
 	rippled.exe!beast::unit_test::suite_info::run(beast::unit_test::runner & r) Line 87	C++
 	rippled.exe!beast::unit_test::runner::run<void>(const beast::unit_test::suite_info & s) Line 188	C++
 	rippled.exe!ripple::test::multi_runner_child::run_multi<ripple::multi_selector>(ripple::multi_selector pred) Line 337	C++
 	rippled.exe!ripple::runUnitTests(const std::string & pattern, const std::string & argument, bool quiet, bool log, bool child, bool ipv6, unsigned __int64 num_jobs, int argc, char * * argv) Line 282	C++
 	rippled.exe!ripple::run(int argc, char * * argv) Line 547	C++
 	rippled.exe!main(int argc, char * * argv) Line 813	C++
 	[External Code]	
```

Reply:
Weird. I put a commit here that prints a bunch of things to stderr that I hope sheds light. Can you please try again? The commit is 4e119da886824c4f91f3512929428b98bf554c7c.

The only thing I can think of is that the unittest calls timerEntry outside of a lock, so there's a possible race condition between setting the phase to establish and setting the result_ in closeLedger(). It's possible I introduced something that triggers the race condition from the standpoint of the test. I know that I'm calling timerEntry() directly now to save time during consensus. The stderr output I added should confirm this is happening.